src/basic/lock-util.c

```diff
@@ -186,3 +189,85 @@ int lock_generic(int fd, LockType type, int operation) {
                 assert_not_reached();
         }
 }
+
+int lock_generic_with_timeout(int fd, LockType type, int operation, usec_t timeout) {
+        _cleanup_(sigkill_waitp) pid_t pid = 0;
+        int r;
+
+        assert(fd >= 0);
+
+        /* A version of lock_generic(), but with a time-out. We do this in a child process, since the kernel
+         * APIs natively don't support a timeout. We set a SIGALRM timer that will kill the child after the
+         * timeout is hit. Returns -ETIMEDOUT if the time-out is hit, and 0 on success. */
```

Comment:
This still confuses me - it takes the lock in the child process, but then waits for it to exit, so when it returns the lock isn't held anymore no? So it just checks that at some point in time nothing else was holding it? What am I missing?

Reply:
BSD locks are per fd. hence the child can take it or the parent, doesn't matter, same effect. after the lock was taken the lock is on the fd, hence all good.

Reply:
Ah so it's a way to have the child take the lock instead of the parent, but it is shared. Please expand the comment to say it, so that those of us that are thicker than average don't spend too much time scratching their heads

Reply:
added a comment now.